src/lib/optimizer/strategy/between_composition_rule.cpp

```diff
@@ -0,0 +1,265 @@
+          break;
+        case PredicateCondition::LessThan:
+          type = ColumnBoundaryType::LowerBoundaryExclusive;
+          break;
+        default:
+          break;
+      }
+    }
+  }
+
+  return {
+      node,
+      column_expression,
+      value_expression,
+      type,
+  };
+}
+
+bool _column_boundary_comparator(const ColumnBoundary& a, const ColumnBoundary& b) {
+  return a.column_expression->column_reference.original_column_id() <
+         b.column_expression->column_reference.original_column_id();
+}
+
+/**
+ * _replace_predicates gets a vector of AbstractLQPNodes as input and
+ * substitutes suitable BinaryPredicateExpressions with BetweenExpressions.
+ * Furthermore BinaryPredicateExpressions which are obsolete after the substitution
+ * are removed.
+**/
+void BetweenCompositionRule::_replace_predicates(std::vector<std::shared_ptr<AbstractLQPNode>>& predicates) const {
+  // Store original input and output
+  auto input = predicates.back()->left_input();
+  const auto outputs = predicates.front()->outputs();
+  const auto input_sides = predicates.front()->get_input_sides();
+
+  auto between_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>();
+  auto predicate_nodes = std::vector<std::shared_ptr<AbstractLQPNode>>();
+  auto boundaries = std::vector<ColumnBoundary>();
+
+  // Filter predicates with a boundary to the boundaries vector
+  for (auto& predicate : predicates) {
+    const auto predicate_node = std::static_pointer_cast<PredicateNode>(predicate);
+    const auto expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate_node->predicate());
+    if (expression != nullptr) {
+      const auto boundary = _get_boundary(expression, predicate_node);
+      if (boundary.type != ColumnBoundaryType::None) {
+        boundaries.push_back(boundary);
+      } else {
+        predicate_nodes.push_back(predicate);
+      }
+    } else {
+      predicate_nodes.push_back(predicate);
+    }
+    // Remove node from lqp in order to rearrange them later
+    lqp_remove_node(predicate);
+  }
+
+  // sort boundaries according to their columns to group boundaries referring to the same column
+  std::sort(boundaries.begin(), boundaries.end(), _column_boundary_comparator);
+
+  // Store the highest lower bound and the lowest upper bound for a column in order to get an optimal BetweenExpression
+  std::shared_ptr<LQPColumnExpression> current_column_expression = nullptr;
+  std::shared_ptr<ValueExpression> lower_bound_value_expression = nullptr;
+  std::shared_ptr<ValueExpression> upper_bound_value_expression = nullptr;
+  auto left_inclusive = true;
+  auto right_inclusive = true;
+
+  // Store all nodes of a column in a temporary node_scope vector.
+  // If no boundary could be created for this column, all nodes in the node_scope have to be recovered.
+  auto node_scope = std::vector<std::shared_ptr<AbstractLQPNode>>();
+
+  for (const auto& boundary : boundaries) {
```

Comment:
I think a normal `column -> constraints` ("constraint" == "ColumnBounds" in your terminology) map would be much simpler than this sorted vector iteration thingy with a lot of temporary state that needs to be "reset" occasionally.